File: backend/app/schemas/surveillance.py
from pydantic import BaseModel
from typing import List, Optional


class SurveillanceKPIs(BaseModel):
    total_liver_cases: int
    total_gi_bleeding_cases: int
    total_hcc_cases: int
    total_procedures: int
    missed_followups: int
    pending_referrals: int
    avg_referral_delay_days: float
    endoscopy_completion_rate: float  # 0-100 %


class DiseaseTrendPoint(BaseModel):
    period: str          # e.g. "2024-W23"
    hep_b: int
    hep_c: int
    cirrhosis: int
    hcc: int
    gi_bleeding: int


class FacilityLoad(BaseModel):
    facility_id: int
    facility_name: str
    region: str
    city: str
    referral_count: int
    procedure_count: int
    liver_case_count: int
    specialist_count: int


class RegionalBurden(BaseModel):
    region: str
    hep_b: int
    hep_c: int
    cirrhosis: int
    hcc: int
    total_cases: int


class ReferralFlowEdge(BaseModel):
    source_facility: str
    target_facility: str
    count: int


class SurveillanceDashboard(BaseModel):
    kpis: SurveillanceKPIs
    disease_trend: List[DiseaseTrendPoint]       # last 12 weeks
    facility_loads: List[FacilityLoad]
    regional_burden: List[RegionalBurden]
    referral_flows: List[ReferralFlowEdge]
    risk_flag_counts: dict                       # NORMAL / OVERDUE / TREND_ALERT


class UserAdminOut(BaseModel):
    id: int
    name: str
    email: str
    role: str
    department: Optional[str] = None
    facility_id: Optional[int] = None
    facility_name: Optional[str] = None
    is_active: bool

    class Config:
        from_attributes = True


class UserRoleUpdate(BaseModel):
    role: str


class UserActiveToggle(BaseModel):
    is_active: bool

File: frontend/lib/api/surveillance.ts
import api from "./client";

export interface SurveillanceKPIs {
  total_liver_cases: number;
  total_gi_bleeding_cases: number;
  total_hcc_cases: number;
  total_procedures: number;
  missed_followups: number;
  pending_referrals: number;
  avg_referral_delay_days: number;
  endoscopy_completion_rate: number;
}

export interface DiseaseTrendPoint {
  period: string;
  hep_b: number;
  hep_c: number;
  cirrhosis: number;
  hcc: number;
  gi_bleeding: number;
}

export interface FacilityLoad {
  facility_id: number;
  facility_name: string;
  region: string;
  city: string;
  referral_count: number;
  procedure_count: number;
  liver_case_count: number;
  specialist_count: number;
}

export interface RegionalBurden {
  region: string;
  hep_b: number;
  hep_c: number;
  cirrhosis: number;
  hcc: number;
  total_cases: number;
}

export interface ReferralFlowEdge {
  source_facility: string;
  target_facility: string;
  count: number;
}

export interface SurveillanceDashboard {
  kpis: SurveillanceKPIs;
  disease_trend: DiseaseTrendPoint[];
  facility_loads: FacilityLoad[];
  regional_burden: RegionalBurden[];
  referral_flows: ReferralFlowEdge[];
  risk_flag_counts: Record<string, number>;
}

export interface UserAdminOut {
  id: number;
  name: string;
  email: string;
  role: string;
  department?: string;
  facility_id?: number;
  facility_name?: string;
  is_active: boolean;
}

export const surveillanceApi = {
  dashboard: (params?: { weeks?: number; facility_id?: number; region?: string }) =>
    api.get<SurveillanceDashboard>("/surveillance/dashboard", { params }),

  listUsers: (params?: { skip?: number; limit?: number }) =>
    api.get<UserAdminOut[]>("/surveillance/users", { params }),

  updateUserRole: (userId: number, role: string) =>
    api.patch<UserAdminOut>(`/surveillance/users/${userId}/role`, { role }),

  toggleUserActive: (userId: number, is_active: boolean) =>
    api.patch<UserAdminOut>(`/surveillance/users/${userId}/active`, { is_active }),

  createUser: (data: {
    name: string;
    email: string;
    password: string;
    role: string;
    department?: string;
    phone?: string;
    facility_id?: number;
  }) => api.post<UserAdminOut>("/admin/users", data),
};
